pwndbg/aglib/shellcode.py

```diff
@@ -0,0 +1,163 @@
+        syscall_bin,
+        restore_context=True,
+        capture=_get_syscall_return_value,
+        disable_breakpoints=disable_breakpoints,
+    )
+
+
+async def exec_shellcode(
+    ec: ExecutionController, blob, restore_context=True, capture=None, disable_breakpoints=False
+):
+    """
+    Tries executing the given blob of machine code in the current context of the
+    inferior, optionally restoring the values of the registers as they were
+    before the shellcode ran, as a means to allow for execution of the inferior
+    to continue uninterrupted. The value of the program counter is always
+    restored.
+
+    Additionally, the caller may specify an object to be called before the
+    context is restored, so that information stored in the registers after the
+    shellcode finishes can be retrieved. The return value of that call will be
+    returned by this function.
+
+    # Safety
+    Seeing as this function injects code directly into the inferior and runs it,
+    the caller must be careful to inject code that will (1) terminate and (2)
+    not cause the inferior to misbehave. Otherwise, it is fairly easy to crash
+    or currupt the memory in the inferior.
+    """
+
+    register_set = pwndbg.lib.regs.reg_sets[pwndbg.aglib.arch.current]
+    preserve_set = register_set.gpr + register_set.args + (register_set.pc, register_set.stack)
+
+    registers = {reg: pwndbg.aglib.regs[reg] for reg in preserve_set}
+    starting_address = registers[register_set.pc]
+
+    # Make sure the blob fits in the rest of the space we have in this page.
+    #
+    # NOTE: Technically, we could actually use anything from the whole page to
+    # all of the pages currently mapped as executable for this. There is no
+    # technical limitation stopping us from doing that, but seeing as doing it
+    # is harder to make sure it works correctly, we don't (for now, at least).
+    page = pwndbg.aglib.vmmap.find(starting_address)
+    assert page is not None
+
+    clearance = page.end - len(blob) - 1
+    if clearance < 0:
+        # The page isn't large enough to hold our shellcode.
+        raise RuntimeError(
+            f"Not enough space to execute code as inferior: \
+            need at least {len(blob)} bytes, have {clearance} bytes available"
+        )
+
+    # Swap the code in the range with our shellcode.
+    existing_code = pwndbg.aglib.memory.read(starting_address, len(blob))
+    pwndbg.aglib.memory.write(starting_address, blob)
+
+    # The continue we use here will trigger an event that would get the context
+    # prompt to show, regardless of the circumstances. We don't want that, so
+    # we preserve the state of the context skip.
+    if pwndbg.dbg.is_gdblib_available():
+        would_skip_context = pwndbg.gdblib.prompt.context_shown
+
+    # Execute.
+    target_address = starting_address + len(blob)
+    bp = pwndbg.dbg.selected_inferior().break_at(BreakpointLocation(target_address), internal=True)
+    while True:
```

Comment:
What is this logic exactly doing? Shouldn't we need to continue until one breakpoint is hit? Why do we need a loop?

How do we need which breakpoint was hit here and if this was the one we were waiting for?

Reply:
> Why do we need a loop?
> How do we know which breakpoint was hit here and if this was the one we were waiting for?

`ExecutionController.cont` raises `CancellationError` if we hit any breakpoints other than the one passed to it as an argument. We use the loop to retry `ec.cont` if it gets cancelled (i.e. it stops before we hit `bp`). This is meant to implement the case where `disable_breakpoints=True` with a behavior that's close to the `gdblib` version of this module[^1].

When `disable_breakpoints=False`, the loop should only ever execute once. Either it hits the breakpoint we expect, or the execution controller gets cancelled and the command finishes running (GDB) or the process runtime returns control to the user (LLDB).

[^1]: The one caveat to this being that right now `ExecutionController` doesn't distinguish between breakpoints from other types of stops, and will lump them all together as "cancellations". That can be fixed fairly easily later by just including that information along with the `CancellationError`.